### PR TITLE
#26804 kubectl run --overrides ignores non existent fields

### PR DESCRIPTION
**Pull request by grpndrs**
thanks @vefimova for his [pr](https://github.com/kubernetes/kubernetes/pull/34771), which was declined due to inactivity for long time.

### CODE REVIEW

**Comment by grpndrs** [comment]
@sttts

File: pkg/kubectl/cmd/util/helpers.go [hunk 1]
```diff
@@ -439,6 +440,9 @@ func Merge(codec runtime.Codec, dst runtime.Object, fragment, kind string) (runt
 	if err != nil {
 		return nil, err
 	}
+	if err = resource.ValidateSchema(patched, schema); err != nil {
+		return nil, fmt.Errorf("error validating resource json after merging with override string: %s\n%v", patched[:], err)
```
Comment (sttts) [inline pkg/kubectl/cmd/util/helpers.go]: this sounds misleading: the printed string is the result of merging, not the override string.

File: pkg/kubectl/cmd/util/helpers.go [hunk 2]
```diff
@@ -439,6 +440,9 @@ func Merge(codec runtime.Codec, dst runtime.Object, fragment, kind string) (runt
 	if err != nil {
 		return nil, err
 	}
+	if err = resource.ValidateSchema(patched, schema); err != nil {
```
Comment (sttts) [inline pkg/kubectl/cmd/util/helpers.go]: This is a no-op for schema=nil ?

File: pkg/kubectl/cmd/run.go [hunk 3]
```diff
@@ -565,7 +566,11 @@ func createGeneratedObject(f cmdutil.Factory, cmd *cobra.Command, generator kube
 
 	if len(overrides) > 0 {
 		codec := runtime.NewCodec(f.JSONEncoder(), f.Decoder(true))
-		obj, err = cmdutil.Merge(codec, obj, overrides, groupVersionKind.Kind)
+		schema, err := f.Validator(cmdutil.GetFlagBool(cmd, "validate"), cmdutil.GetFlagString(cmd, "schema-cache-dir"))
```
Comment (sttts) [inline pkg/kubectl/cmd/run.go]: Shouldn't `"schema-cache-dir"` be abstracted somehow? I guess this string will be elsewhere as well.

**Comment by sttts** [comment]
A test-cmd.sh test is missing.

**Comment by grpndrs** [comment]
Thanks @sttts Looking into the shortcomings

File: pkg/kubectl/cmd/run.go (hunk 3 above)
Comment (grpndrs) [inline pkg/kubectl/cmd/run.go]: Sorry, I am unable to understand your point here. I checked `"schema-cache-dir"` usage at other places in the code and found it similar to this one.

File: pkg/kubectl/cmd/util/helpers.go (hunk 2 above)
Comment (grpndrs) [inline pkg/kubectl/cmd/util/helpers.go]: @sttts So, should we have something like
`if schema == nil {if err = resource.ValidateSchema(patched, schema); err != nil {...}
 }` ?

File: pkg/kubectl/cmd/util/helpers.go (hunk 1 above)
Comment (grpndrs) [inline pkg/kubectl/cmd/util/helpers.go]: @sttts We can remove "override string" part. Thanks for the review.

**Comment by grpndrs** [comment]
> A test-cmd.sh test is missing.

@sttts Should we add test in this file?

**Comment by sttts** [comment]
> A test-cmd.sh test is missing.
> @sttts Should we add test in this file?

If it's feasible, I would like to see such a test. They are the e2e tests of kubectl.

**Comment by spiffxp** [comment]
@k8s-bot ok to test
pinging @kubernetes/sig-cli-pr-reviews

**Pull request closed without merging**